addons/stock_picking_batch/report/report_picking_batch.xml

```diff
@@ -70,55 +70,58 @@
-                                            </td>
-                                        </tr>
-                                        <tr t-foreach="loc_move_line" t-as="move_operation">
-                                            <td>
-                                                <span t-field="move_operation.display_name"/>
-                                            </td>
-                                            <td>
-                                                <t t-if="not has_package">
-                                                    <t t-if="move_operation.state == 'done'">
+                                <t t-foreach="locations" t-as="location_from">
+                                    <t t-set="loc_move_line" t-value="move_line_ids.filtered(lambda x: x.location_id==location_from)"/>
+                                    <t t-foreach="loc_move_line.location_dest_id" t-as="location_dest">
+                                        <t t-set="move_lines" t-value="loc_move_line.filtered(lambda x: x.location_dest_id==location_dest)"/>
+                                        <t t-set="products" t-value="move_lines.product_id"/>
+                                        <tbody>
+                                            <tr>
+                                                <td style="background-color:lightgrey" colspan="6" >
+                                                    <p style="margin:0px"><strong>FROM</strong>
+                                                    <span t-esc="move_lines.location_id.display_name"/></p>
+                                                    <p style="margin:0px"><strong>TO</strong>
+                                                    <span t-esc="move_lines.location_dest_id.display_name"/></p>
+                                                </td>
+                                            </tr>
+                                            <tr t-foreach="move_lines" t-as="move_operation">
+                                                <td>
+                                                    <span t-field="move_operation.display_name"/>
+                                                </td>
+                                                <td>
+                                                    <t t-if="not has_package">
+                                                        <t t-if="move_operation.state == 'done'">
+                                                            <span t-esc="sum(move_operation.mapped('qty_done'))"/>
+                                                        </t>
+                                                        <t t-else="">
+                                                            <span t-esc="sum(move_operation.mapped('product_uom_qty'))"/>
+                                                        </t>
+                                                    </t>
+                                                    <t t-if="has_package">
```

Comment:
Not related to the issue, but since you modify these lines, maybe you can take advantage of it to shorter the `has_package`/`not has_package` part ?
```xml
                                                    <span t-if="has_package or move_operation.state == 'done'"
                                                          t-esc="sum(move_operation.mapped('qty_done'))"/>
                                                    <span t-else="" t-esc="sum(move_operation.mapped('product_uom_qty'))"/>
```
It allows to reduce 10 lines to only 3 :grin:

Reply:
Indeed, it's cleaner ! Thanks 😉